recipe-server/normandy/recipes/models.py

```diff
@@ -373,6 +393,52 @@ def save(self, *args, **kwargs):
         self.implementation_hash = self.compute_implementation_hash()
         super().save(update_fields=['implementation_hash'])
 
+    def validate_arguments(self, arguments, old_arguments=None):
+        """
+        Test if `arguments` follows all action-specific rules.
+
+        If there is a previous version of the arguments, they should
+        be passed as `old_arguments`. Some validation rules depend on
+        the history of the arguments.
```

Comment:
We can't use the dirty mixin because `Recipe` doesn't store any data. All the data is in revisions, which don't change over time. Given that, it made more sense for `Action` instances to validate their own arguments.